### PR TITLE
ZCS-1200:Need way to query server for total number of GAL entries

### PR DESCRIPTION
**Pull request by shrikant-prasad**
added optional attribute 'getCount' in SyncGalRequest. response will have 'remain' attribute when getCount= true and it will contain count of remaining items".

### CODE REVIEW

File: store/src/java/com/zimbra/cs/db/DbMailItem.java [hunk 1]
```diff
@@ -2778,6 +2779,54 @@ public static UnderlyingData getByHash(Mailbox mbox, String hash) throws Service
         }
     }
 
+    public static int getModifiedItemsCount(Mailbox mbox, MailItem.Type type, long lastSync,
+            int sinceDate, Set<Integer> visible, int lastDeleteSync)
```
Comment (chinmay15) [inline store/src/java/com/zimbra/cs/db/DbMailItem.java]: remove unused param lastDeleteSync

File: store/src/java/com/zimbra/cs/db/DbMailItem.java [hunk 2]
```diff
@@ -2778,6 +2779,54 @@ public static UnderlyingData getByHash(Mailbox mbox, String hash) throws Service
...
+    public static int getModifiedItemsCount(Mailbox mbox, MailItem.Type type, long lastSync,
+            int sinceDate, Set<Integer> visible)
+    throws ServiceException {
+        if (Mailbox.isCachedType(type)) {
+            throw ServiceException.INVALID_REQUEST("folders and tags must be retrieved from cache", null);
+        }
+        DbConnection conn = mbox.getOperationConnection();
+        PreparedStatement stmt = null;
+        try {
+            String typeConstraint = type == MailItem.Type.UNKNOWN ? "type NOT IN " + NON_SYNCABLE_TYPES : typeIn(type);
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/db/DbMailItem.java]: There are a lot of intermediate Strings being created here to create the statement.  How about sticking with `StringBuilder` syntax and doing something like:

    buf.append("SELECT COUNT(id)")
        .append(" FROM ")
        .append(getMailItemTableName(mbox))
        .append(" WHERE ")
        .append(IN_THIS_MAILBOX_AND)
        .append("mod_metadata > ? AND ");
        .append(sinceDate > 0 ? "date > ? AND " : "")
        .append(MailItem.Type.UNKNOWN ? "type NOT IN " + NON_SYNCABLE_TYPES : typeIn(type))
    if (visible != null && visible.size() > 0) {
        String folderString = visible.stream().map(i -> i.toString()).collect(Collectors.joining(", "));
        buf.append(" AND folder_id IN ( ").append(folderString).append(" )");
    }

File: soap/src/java/com/zimbra/soap/account/message/SyncGalRequest.java [hunk 3]
```diff
@@ -63,6 +63,13 @@
     @XmlAttribute(name=AccountConstants.A_ID_ONLY /* idOnly */, required=false)
     private ZmBoolean idOnly;
 
+    /**
+     * @zm-api-field-tag getCount
+     * @zm-api-field-description Flag whether count of remaining records should be returned in response or not.
```
[the remark this answers is not in the document]
Comment (shrikant-prasad) [reply soap/src/java/com/zimbra/soap/account/message/SyncGalRequest.java]: Done

File: store/src/java/com/zimbra/cs/db/DbMailItem.java (hunk 2 above)
Comment (shrikant-prasad) [reply store/src/java/com/zimbra/cs/db/DbMailItem.java]: Done

**Pull request closed without merging**